### PR TITLE
Update readme around end to end testing

### PR DESCRIPTION
**Pull request by peternied**
### Description
Update readme around end to end testing

### Check List
- [ ] ~New functionality includes testing~
  - [ ] ~All tests pass, including unit test, integration test and doctest~
- [X] New functionality has been documented
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -59,12 +69,71 @@ The FetchMigration directory hosts tools that simplify the process of backfillin
...
+        MC --> SC
+        MC --> TC
+    end
+```
+
+#### Local end to end testing
+
+Locally performing end to end test cases is done slightly differently, by subtituting cloud services for the local file system, and TestContainers to provider source and target migration clusters its much faster to validate different configurations and versions.  By running inside of a JUnit JVM this also allows deep access into the running historical worker for monitoring or fault injection.  When test cases themselves do not require full source or target clusters using mock endpoints or data can greatly accelerate test speed.
+
+These tests are all be run with the command `./gradlew test` from the root directory.
```
Comment (peternied) [inline README.md:113]: Hmm, what about some of the python code we have, seems like we are using the python test runner - does that apply here or is do we think that is basically covered in the component/unit test kind of area.

File: README.md [hunk 2]
```diff
@@ -59,12 +69,71 @@ The FetchMigration directory hosts tools that simplify the process of backfillin
...
+
+        MC --> ECS
+        MC --> SC
+        MC --> TC
+    end
+```
+
+#### Local end to end testing
+
+Locally performing end to end test cases is done slightly differently, by subtituting cloud services for the local file system, and TestContainers to provider source and target migration clusters its much faster to validate different configurations and versions.  By running inside of a JUnit JVM this also allows deep access into the running historical worker for monitoring or fault injection.  When test cases themselves do not require full source or target clusters using mock endpoints or data can greatly accelerate test speed.
```
Comment (mikaylathompson) [inline README.md:111]: nit: subtituting -> substituting

Comment (mikaylathompson) [reply README.md:111]: and "**using** Test Containers to **provide** source and target migration clusters**, it's**"

File: README.md
Comment (mikaylathompson) [inline README.md:1]: Do you see a role for docker-level local end to end tests? In particular, we have generally prioritized making the docker/opensource tooling an equal citizen with the AWS-specific tooling. In this document, it looks like the docker solution is not explicitly tested (my read of this is that the local testing is effectively testing, e.g. RFS, using containers in a supporting role, not thinking of it as testing the docker solution as an e2e system)

File: README.md (hunk 1 above)
Comment (mikaylathompson) [reply README.md:113]: Currently, we use python tests both for unit testing (e.g. console lib or fetch), as well as to run our existing end to end tests. It might be possible to run the latter from within gradle (though I'm not sure we gain much from that), but my understanding from this diagram (and the emphasis on running the tests in the JVM) is that you're suggesting rewriting/porting/scrapping those tests. Is that accurate? If so, running python tests might not be that much of a concern.

Is that what you're asking here, or are you wondering if there's a role that we're using python for that java tests couldn't fill?

File: README.md (hunk 2 above)
Comment (peternied) [reply README.md:111]: Thank you! Btw I found out I wasn't using a spell checker in my IDE - fixed!

File: README.md (hunk 1 above)
Comment (peternied) [reply README.md:113]: Good distinction, python is well suited for the defining the console link CLI and API.  This code is hard to test beyond the unit level with anything other than the real control plane systems (be it a cloud service, kuberneties, or local docker) and java-based migration tools.  What do you think of this?

I guess I'm saying that testing the python code in an end to end way involves `Cloud end to end testing` kind of structure rather than another category - I am curious we think there is a way to test the python codebase that is missing here. (Fake java migration tools that spit out errors?)

File: README.md
Comment (peternied) [reply README.md:1]: @mikaylathompson  There is a huge amount of utility for local testing to be up in docker, my concern is about how much time we'd spend getting Docker to work instead of Kubernetes which handles abstractions a cleaner level.  Its an interesting problem because today we are using the docker based end-to-end testing as an entry before the 'cloud' testing in jenkins ci stack.

Maybe you've got an idea for a section that describes what/how docker fits in, want to write something up and submit a PR to my repo or just make a comment here?

Comment (mikaylathompson) [reply README.md:1]: I'm grouping Docker and Kubernetes together as the open source solution. I'm fine with postponing the actual testing until we've stabilized that solution, but I don't see any mention of how either of those fits into this e2e testing framework.

File: README.md [hunk 3]
```diff
@@ -59,12 +69,71 @@ The FetchMigration directory hosts tools that simplify the process of backfillin
...
 
-More documentation on this test script can be found here:
-[End-to-End Testing](test/README.md)
+#### Component and Unit testing
+
+While end to end tests can be very useful for testing integration boundaries, for simplier components and functions it is advised to write junit tests that focus on edge cases.  When adding functionality its advised to start at these lower level test and then see what end to end tests might need to be included or modified to check for coverage if needed.
+
+#### Cloud end to end testing
+
+These tests are setup once with a deployment into a cloud ecosystem, because full infrastructure is inheriantly expensive monitarily and time wise - there will be only a small number of tests that make sure the migrations tools are working as expected.  In this example diagram, tests to validate historical migration from a snapshot use the full cloud resources.  Control of the test scenarios is largely through the deployed migration console.
```
Comment (mikaylathompson) [inline README.md:82]: nits: inheriantly -> inherently; monitarily -> monetarily;

File: README.md
Comment (mikaylathompson) [reply README.md:1]: It seems to fit in between the cloud and local e2e solutions as you've laid them out. It _could_ be done locally, but doesn't really (I think) make sense to do it as part of the JVM. If you're thinking of it as a cloud thing, that's fine with me, but I think it deserves to be called out somwehere.

File: README.md (hunk 1 above)
Comment (mikaylathompson) [reply README.md:113]: Yep, I'm fine with that. I think we definitely could (should?) do some testing of the CLI and API that involves fake/toy containers running locally , but I don't think it's a particularly high priority to make that a local thing instead of a cloud-level thing.

**Comment by peternied** [comment]
@mikaylathompson I've been thinking about where we go with this PR and what it means to merge it.  While it is useful to call out these kinds of ideals, as you mention the gap in 'local' vs 'cloud' testing where we might use docker / kubernetes is pretty big.

I'm going to close out this PR as it doesn't really change anything about our existing systems.  As we look to do the 'grand' refactor to make this project more straight forward we can revisit this topic.
